File: kernel/trace/ifr.c
/* kernel/trace/ifr.c */
#include "ifr.h"
#include "../time/time.h"
#include "../lib/sha256.h"
#include "../ipc/heap.h"
#include "flightrec.h"
#include <string.h>

void ifr_build(ifr_record_t *out,
               uint32_t job_id,
               uint32_t episode_id,
               uint32_t model_id,
               const float *profile,
               uint16_t profile_len,
               float goodput) {
    if (!out)
        return;

    memset(out, 0, sizeof(*out));
    out->magic = IFR_MAGIC;
    out->version = IFR_VERSION_V2;
    out->flags = 0;
    out->job_id = job_id;
    out->episode_id = episode_id;
    out->model_id = model_id;
    out->record_size = sizeof(*out);
    out->ts_usec = time_usec();
    out->goodput = goodput;

    if (profile && profile_len > 0) {
        if (profile_len > IFR_PROFILE_MAX)
            profile_len = IFR_PROFILE_MAX;
        out->profile_len = profile_len;
        for (uint16_t i = 0; i < profile_len; i++) {
            out->profile[i] = profile[i];
        }
    }

    sha256_hash((const uint8_t *)out, offsetof(ifr_record_t, hash), out->hash);
}

int ifr_verify(const ifr_record_t *rec) {
    if (!rec)
        return 0;
    if (rec->magic != IFR_MAGIC)
        return 0;
    if (rec->version != IFR_VERSION_V2)
        return 0;
    if (rec->record_size != sizeof(*rec))
        return 0;
    if (rec->profile_len > IFR_PROFILE_MAX)
        return 0;

    uint8_t expected[32];
    sha256_hash((const uint8_t *)rec, offsetof(ifr_record_t, hash), expected);
    return (memcmp(expected, rec->hash, sizeof(expected)) == 0) ? 1 : 0;
}

static void fill_nonce(uint8_t nonce[32]) {
    sha256_ctx_t ctx;
    sha256_init(&ctx);
    uint64_t t = time_usec();
    uint64_t c = time_cycles();
    sha256_update(&ctx, (const uint8_t *)&t, sizeof(t));
    sha256_update(&ctx, (const uint8_t *)&c, sizeof(c));
    sha256_final(&ctx, nonce);
}

static void compute_model_digest(uint32_t model_id, uint8_t out[32], uint16_t *flags) {
    void *data = heap_get_data((uint16_t)model_id);
    uint32_t size = heap_get_blob_size((uint16_t)model_id);
    if (!data || size == 0) {
        memset(out, 0, 32);
        if (flags) {
            *flags |= IFR_FLAG_MODEL_DIGEST_MISSING;
        }
        return;
    }
    sha256_hash((const uint8_t *)data, size, out);
}

static void compute_policy_digest(uint8_t out[32], uint16_t *flags) {
    const char *policy = "zenedge-policy-v1";
    sha256_hash((const uint8_t *)policy, strlen(policy), out);
    if (flags) {
        *flags |= IFR_FLAG_POLICY_DIGEST_PLACEHOLDER;
    }
}

void ifr_build_v3(ifr_record_v3_t *out,
                  const uint8_t prev_chain_hash[32],
                  uint32_t job_id,
                  uint32_t episode_id,
                  uint32_t model_id,
                  float goodput) {
    if (!out)
        return;

    memset(out, 0, sizeof(*out));
    out->magic = IFR_MAGIC;
    out->version = IFR_VERSION_V3;
    out->flags = 0;
    out->record_size = sizeof(*out);
    out->job_id = job_id;
    out->episode_id = episode_id;
    out->model_id = model_id;
    out->ts_usec = time_usec();
    out->goodput = goodput;

    fill_nonce(out->nonce);
    compute_model_digest(model_id, out->model_digest, &out->flags);
    compute_policy_digest(out->policy_digest, &out->flags);

    if (prev_chain_hash) {
        memcpy(out->prev_chain_hash, prev_chain_hash, 32);
    }

    flightrec_seal_hash(out->flightrec_seal_hash);

    /* IFR core hash excludes ifr_hash, chain_hash, and signature fields. */
    sha256_hash((const uint8_t *)out, offsetof(ifr_record_v3_t, ifr_hash), out->ifr_hash);

    sha256_ctx_t ctx;
    sha256_init(&ctx);
    sha256_update(&ctx, out->prev_chain_hash, 32);
    sha256_update(&ctx, out->ifr_hash, 32);
    sha256_update(&ctx, out->flightrec_seal_hash, 32);
    sha256_update(&ctx, out->nonce, 32);
    sha256_update(&ctx, out->model_digest, 32);
    sha256_update(&ctx, out->policy_digest, 32);
    sha256_final(&ctx, out->chain_hash);

    memset(out->sig_classical, 0, sizeof(out->sig_classical));
    out->flags |= IFR_FLAG_SIG_UNAVAILABLE;
}

int ifr_verify_v3(const ifr_record_v3_t *rec) {
    if (!rec)
        return 0;
    if (rec->magic != IFR_MAGIC)
        return 0;
    if (rec->version != IFR_VERSION_V3)
        return 0;
    if (rec->record_size != sizeof(*rec))
        return 0;

    uint8_t expected_ifr[32];
    sha256_hash((const uint8_t *)rec, offsetof(ifr_record_v3_t, ifr_hash), expected_ifr);
    if (memcmp(expected_ifr, rec->ifr_hash, 32) != 0)
        return 0;

    sha256_ctx_t ctx;
    sha256_init(&ctx);
    sha256_update(&ctx, rec->prev_chain_hash, 32);
    sha256_update(&ctx, rec->ifr_hash, 32);
    sha256_update(&ctx, rec->flightrec_seal_hash, 32);
    sha256_update(&ctx, rec->nonce, 32);
    sha256_update(&ctx, rec->model_digest, 32);
    sha256_update(&ctx, rec->policy_digest, 32);
    uint8_t expected_chain[32];
    sha256_final(&ctx, expected_chain);

    return (memcmp(expected_chain, rec->chain_hash, 32) == 0) ? 1 : 0;
}

File: bridge/ifr.py
"""
IFR parsing and verification helpers.
"""

import hashlib
from typing import Optional, Dict, Any

from .protocol import (
    IFR_V2_STRUCT,
    IFR_V3_STRUCT,
    IFR_MAGIC,
    IFR_VERSION_V2,
    IFR_VERSION_V3,
    IFR_PROFILE_MAX,
    IFR_V2_HASH_OFFSET,
    IFR_V3_HASH_OFFSET,
)


def parse_ifr_blob(data: bytes) -> Optional[Dict[str, Any]]:
    if not data or len(data) < 8:
        return None

    magic = int.from_bytes(data[0:4], "little")
    version = int.from_bytes(data[4:6], "little")

    if magic != IFR_MAGIC:
        return None

    if version == IFR_VERSION_V2 and len(data) >= IFR_V2_STRUCT.size:
        unpacked = IFR_V2_STRUCT.unpack(data[:IFR_V2_STRUCT.size])
        (magic, version, flags, job_id, episode_id, model_id, record_size, ts_usec,
         goodput, profile_len, _reserved, *rest) = unpacked
        profile = rest[:IFR_PROFILE_MAX]
        hash_bytes = rest[IFR_PROFILE_MAX]

        if record_size != IFR_V2_STRUCT.size or profile_len > IFR_PROFILE_MAX:
            return None

        expected = hashlib.sha256(data[:IFR_V2_HASH_OFFSET]).digest()
        hash_ok = (expected == hash_bytes)

        return {
            "magic": magic,
            "version": version,
            "flags": flags,
            "job_id": job_id,
            "episode_id": episode_id,
            "model_id": model_id,
            "record_size": record_size,
            "ts_usec": ts_usec,
            "goodput": goodput,
            "profile_len": profile_len,
            "profile": list(profile[:profile_len]),
            "hash": hash_bytes,
            "hash_ok": hash_ok,
        }

    if version == IFR_VERSION_V3 and len(data) >= IFR_V3_STRUCT.size:
        unpacked = IFR_V3_STRUCT.unpack(data[:IFR_V3_STRUCT.size])
        (magic, version, flags, record_size, job_id, episode_id, model_id,
         ts_usec, goodput, nonce, model_digest, policy_digest,
         flightrec_seal_hash, prev_chain_hash, ifr_hash, chain_hash,
         sig_classical) = unpacked

        if record_size != IFR_V3_STRUCT.size:
            return None

        expected_ifr = hashlib.sha256(data[:IFR_V3_HASH_OFFSET]).digest()
        ifr_ok = (expected_ifr == ifr_hash)

        chain_ctx = hashlib.sha256()
        chain_ctx.update(prev_chain_hash)
        chain_ctx.update(ifr_hash)
        chain_ctx.update(flightrec_seal_hash)
        chain_ctx.update(nonce)
        chain_ctx.update(model_digest)
        chain_ctx.update(policy_digest)
        expected_chain = chain_ctx.digest()
        chain_ok = (expected_chain == chain_hash)

        return {
            "magic": magic,
            "version": version,
            "flags": flags,
            "job_id": job_id,
            "episode_id": episode_id,
            "model_id": model_id,
            "record_size": record_size,
            "ts_usec": ts_usec,
            "goodput": goodput,
            "nonce": nonce,
            "model_digest": model_digest,
            "policy_digest": policy_digest,
            "flightrec_seal_hash": flightrec_seal_hash,
            "prev_chain_hash": prev_chain_hash,
            "ifr_hash": ifr_hash,
            "chain_hash": chain_hash,
            "sig_classical": sig_classical,
            "hash_ok": ifr_ok and chain_ok,
            "chain_ok": chain_ok,
        }

    return None
